Bottom: 84bbea7c0e115f152706579f56b36b61dea90111
Top:    aa8f70a5a4e7bc141fd2ab499718aa266dfa2afa
Author: Jon Smirl <jonsmirl@gmail.com>
Date:   2014-07-10 12:03:38 -0400

Refresh of audio

---

diff --git a/drivers/dma/sun4i-dma.c b/drivers/dma/sun4i-dma.c
index db6579d..ca33f8c 100644
--- a/drivers/dma/sun4i-dma.c
+++ b/drivers/dma/sun4i-dma.c
@@ -523,6 +523,7 @@ static void sun4i_dma_free_contract(struct virt_dma_desc *vd)
 	struct sun4i_dma_contract *contract = to_sun4i_dma_contract(vd);
 	struct sun4i_dma_promise *promise;
 
+	printk("JDS DMA sun4i_dma_free_contract\n");
 	/* Free all the demands and completed demands */
 	list_for_each_entry(promise, &contract->demands, list) {
 		kfree(promise);
@@ -937,21 +938,22 @@ static irqreturn_t sun4i_dma_interrupt(int irq, void *dev_id)
 		if (bit & 1) {
 			spin_lock(&vchan->vc.lock);
 			if (contract->cyclic) {
-				vchan_cyclic_callback(&contract->vd);
-
 				/* move promise to back of list */
-				list_del(&vchan->processing->list);
+				list_del_init(&vchan->processing->list);
 				list_add_tail(&vchan->processing->list, &contract->demands);
+				vchan->processing = NULL;
+
+				vchan_cyclic_callback(&contract->vd);
 			} else {
 				/*
 				 * Move the promise into the completed list now that
 				 * we're done with it
 				 */
-				list_del(&vchan->processing->list);
+				list_del_init(&vchan->processing->list);
 				list_add_tail(&vchan->processing->list, &contract->completed_demands);
 				vchan->pchan = NULL;
+				vchan->processing = NULL;
 			}
-			vchan->processing = NULL;
 			spin_unlock(&vchan->vc.lock);
 
 			irqs &= ~BIT(bit);
